### PR TITLE
Add support for multiple attachments

### PR DESCRIPTION
**Pull request by ecridge**
This changeset allows responses to be checked for multiple attachments, sensitive to their number, order, and type. It also allows for the text body sent along with the attachments to be (optionally) checked just like it is for a plain response, and makes it trivial to add support for more attachment types in the future.

The YAML syntax is unchanged (save for the addition of an `<OTHER>` attachment type as a catch-all), and these changes should be fully backwards compatible. To achieve this, I made two compromises, which I think are fairly natural for users: (1) attachments must come at the end of the message spec; and (2) the text body of a response is not checked if the message spec consists _only_ of attachments.

### CODE REVIEW

**Comment by ecridge** [comment]
@lingz one area I could use your input on is how to make sure that the output is still friendly when a response fails to match

**Comment by lingz** [comment]
Hey Joe,

Thanks a lot for this, I will do a full review tomorrow but cursory inspection looks good. A quick idea I have would be,

```
Attachment mismatch:
Expected: 2 Images, 1 Button
Got: No Attachments
```

Something like this might work?

File: README.md [hunk 1]
```diff
@@ -71,19 +71,39 @@ Dialogue:
...
-`<CARDS>` | A card attachment
 `<(REGEX)>` | Any regex expression, i.e. `<([0-9]{2})>`
 `<$VARNAME>` | An expression from the locale/translation file
 
+You can match one or more attachments **on their own, or at the end** of the
+response:
+
+Tag | Meaning
+--- | ---
+`<CARDS>` | Any card attachment
```
Comment (lingz) [inline README.md]: Thanks for the docs here. I would just add a comment here that <CARDS> includes buttons, as this might not be obvious to them.

File: README.md [hunk 2]
```diff
@@ -71,19 +71,39 @@ Dialogue:
...
+`<IMAGE>` | Any image attachment
+`<OTHER>` | Any other type of attachment
```
Comment (lingz) [inline README.md]: Add some examples below here.

File: src/runner.ts [hunk 3]
```diff
@@ -156,7 +156,7 @@ export class Runner {
       if (response !== null) {
         if (program.verbose) {
           const timing = (new Date().getTime() - this.timing) || 0;
-          if (response.messageTypes.includes(MessageType.Text)) {
+          if (response.attachments.length === 0) {
```
Comment (lingz) [inline src/runner.ts]: This seems wrong? Should be `if (response.text)`. Also, for printing, we should probably print the attachments if they exist like:

```
Bot: Here you go: <IMAGE> <BUTTON> <BUTTON>
```

I know this is not how it is before but this would bring it into consistency with the new formatting style.

File: src/spec/response.ts [hunk 4]
```diff
@@ -1,50 +1,80 @@
...
-        break;
-      case '<CARDS>':
-        this.responseType = MessageType.Card;
-        break;
-      default:
-        this.responseType = MessageType.Text;
-        this.textMatchChecker = new RegExp(Response.transformTags(this.original));
+    this.bodyText = this.original;
+
+    // Construct a list of attachments from the attachment wildcards at the
```
Comment (lingz) [inline src/spec/response.ts]: I like this solution, it's nice. Though I wonder why you constrain it to happening at the end of the line? Seems a bit artificial to me?

File: src/spec/response.ts [hunk 5]
```diff
@@ -1,50 +1,80 @@
...
+    if (this.bodyText && message.text.trim().match(this.textMatchChecker) === null) {
+      return false;
+    }
+
+    // Check that we have the same number, kind, and order of attachments.
+    // We can use JSON.stringify here since we know attachments is just a
+    // simple array of strings.
+
+    // Check that we have the right number of attachments.
+    if (JSON.stringify(this.attachments) !== JSON.stringify(message.attachments)) {
```
Comment (lingz) [inline src/spec/response.ts]: Stringify has this tricky behavior where ordering of keys is only guaranteed in Node environment **IF** the object is the same (same pointer). Even if two objects look the same, they will not necessarily get stringified in the same way. 

Use https://github.com/substack/json-stable-stringify to do a sort before the stringify.

File: src/test/spec/response.ts [hunk 6]
```diff
@@ -19,29 +19,94 @@ describe('response.ts', () => {
...
   });
   it('should match <CARDS> correctly', () => {
     expect(new Response(' <CARDS> ').matches({
       user: null,
       text: null,
-      messageTypes: [MessageType.Card],
+      attachments: [Attachment.Cards],
     })).to.be.true;
   });
+  it('should ignore text with <IMAGE> if no text is expected', () => {
```
Comment (lingz) [inline src/test/spec/response.ts]: Actually I think this should be a breaking change, and we can increment the major version of autobot to suggest this, and put a note in README.md

File: src/test/spec/response.ts [hunk 7]
```diff
@@ -19,29 +19,94 @@ describe('response.ts', () => {
...
+    // This is required for backwards compatibility with when <IMAGE>
+    // could only be used as the entire response.
+    expect(new Response(' <IMAGE> ').matches({
+      user: null,
+      text: ' Here is your image: ',
+      attachments: [Attachment.Image],
+    })).to.be.true;
+  });
+  it('should ignore text with <CARDS> if no text is expected', () => {
```
Comment (lingz) [inline src/test/spec/response.ts]: Same here, so I would reverse the test condition, and make it `.to.be.false`

**Review by lingz — CHANGES_REQUESTED** [review]
Joe this looks really nice to me, apart from small comments, the only suggestion would be to make the breaking change now to prevent technical debt accumulation, since the number of users is relatively low. I'll increment the major version number to let people know, but if you add the note to README.md, that would be great. If you fix these, I can merge this and release on npm.

File: src/spec/response.ts (hunk 4 above)
Comment (ecridge) [reply src/spec/response.ts]: In principle they could go anywhere, but I’m not sure what the use case would be. For example:

```YAML
Dialogue:
  - Human: How is my campaign doing?
  - Bot: According to <IMAGE> and <IMAGE>, it’s going great!
```

This doesn’t make any sense in bot framework _per se_, because it doesn’t define a mechanism for embedding attachments inline with the text. Presumably you would expect `attachments` to be `[Attachment.Image, Attachment.Image]`, but is `text` supposed to be `'According to'`, or `'According to and , it’s going great!`, or something else?

You can imagine a client which does support this (e.g. one that uses `$0`, `$1`, … to reference attachments in the `attachments` array), in which case you could check that `text === 'According to $0 and $1, it’s going great!'`, and that `attachments[0] === Attachment.Image` and `attachments[1] === Attachment.Image`; but in the absence of an established convention I’m not sure how useful this would be.

File: src/spec/response.ts (hunk 5 above)
Comment (ecridge) [reply src/spec/response.ts]: Interesting. I assumed that stringifying an Array\<String\> would be safe. If you like I can just revert ac0dd3b to avoid bringing in another library?

File: src/runner.ts (hunk 3 above)
Comment (ecridge) [reply src/runner.ts]: Ah good catch, thanks.

File: src/spec/response.ts (hunk 5 above)
Comment (lingz) [reply src/spec/response.ts]: Yeah that sounds like a good idea

File: src/spec/response.ts (hunk 4 above)
Comment (lingz) [reply src/spec/response.ts]: Actually you are right that the botbuilder platform specifically limits this ordering. However, for other platforms like Slack, it is allowed to have intermingled text + attachments, so for that reason it might be good to allow the tags anywhere, and just concatenate the text.

It could lead to this edge case where someone sends

```
Hi <IMAGE> there <IMAGE>
```

and it validates against:

```
HI there <IMAGE> <IMAGE>
```

However at the same time, we could introduce this fix as a separate PR as you don't have any example connector right now, so it would be hard to test. For now we can concatenate and it will work for most practical use cases I think.

So to summarize I think for now a good compromise is to just drop the ordering condition (remove the `$` on the regex parser).
